### PR TITLE
Improves information about container and scopes

### LINKED ISSUE #494: Container with isolated memory scopes
https://github.com/spiral/framework/pull/870

### PR DESCRIPTION
**Pull request by butschster**
#494

### CODE REVIEW

File: docs/en/framework/scopes.md [hunk 1]
```diff
@@ -1,50 +1,208 @@
...
+
+You can optionally give a name to the scope with the `$name` parameter. This can be useful for tracking or managing
+specific scopes.
+
+**Restrictions**
+
+These restrictions ensure proper usage and prevent conflicts within the scope hierarchy. Here are the key restrictions:
+
+1. **Named Scope Uniqueness:** Named scopes within the same hierarchy must have unique names. This means that you cannot
+   have multiple scopes with the same name at the same level of the scope hierarchy. Each named scope should have a
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]: > you cannot have multiple scopes with the same name at the same level of the scope hierarchy

Looks wrong. Parallel scopes are allowed (see №4)

File: docs/en/framework/scopes.md [hunk 2]
```diff
@@ -1,50 +1,208 @@
...
+   distinct and meaningful name to avoid any ambiguity.
+
+2. **Default Bindings for Named Scopes:** Named scopes can have default bindings associated with them. It's important to
+   note that default bindings do not affect already created container instances, except for the root container. Default
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]:
```suggestion
   note that changing of default bindings does not affect already created container instances with the same scope name, except for the root container. Default
```

Comment (roxblnfk) [reply docs/en/framework/scopes.md]: I think the restriction is not a restriction. It's a behavior of the container binder.

File: docs/en/framework/scopes.md [hunk 3]
```diff
@@ -1,50 +1,208 @@
...
+   bindings provide a convenient way to set up common dependencies for a particular named scope.
+
+3. **Root Scope Name:** The most parent scope in the hierarchy is named the root scope by default. The root scope serves
+   as the top-level scope and provides the foundation for all other scopes. It's the starting point for dependency
+   resolution and can have its own set of default bindings.
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md:180]: The meaning is like this:

> The most parent scope has the name `root` by default.

It's also doesn't look like a restriction.

File: docs/en/framework/scopes.md [hunk 4]
```diff
@@ -1,50 +1,208 @@
...
+
+4. **Parallel Named Scopes:** Parallel named scopes are allowed within the hierarchy. This means that you can create
+   multiple named scopes at the same level, as long as their names are unique.
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]:
```suggestion
4. **Parallel Named Scopes:** Parallel named scopes are allowed.
```

File: docs/en/framework/scopes.md [hunk 5]
```diff
@@ -1,50 +1,208 @@
...
+
+### Resolving Rules
+
+When a nested scope does not have its own bindings, it will now resolve dependencies from the parent scope. This allows
+for more flexible dependency resolution, with dependencies being resolved in the closest parent scope that has the
+required bindings.
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]: Need to fix the second sentence

File: docs/en/framework/scopes.md [hunk 6]
```diff
@@ -1,50 +1,208 @@
...
+
+### Accessing Scoped Values
+
+You can access the values set in the scope directly from the container or via method injection in your services or
+controllers. However, this should only be done within the IoC scope.
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]:
```suggestion
You can access the values set in the scope directly from the container or via dependency injection in your services or
controllers. However, this should only be done within the IoC scope.
```

Btw is the sentence required?

File: docs/en/framework/scopes.md [hunk 7]
```diff
@@ -1,50 +1,208 @@
...
-## Explanation
+## How It Works
 
-While processing some user request, the context-specific data is located in the IoC scope, available in the container
-only for a limited time. Such operation is performed using `Spiral\Core\ScopeInterface`->`runScope` method. Default
-spiral container `Spiral\Core\Container` implements this interface.
+When processing a user request, the relevant data for that request is placed in a scoped area of the IoC container. This
+data is only available for a limited time while processing that request.
+
+You can use the `Spiral\Core\ScopeInterface->runScope` method to achieve this. The default Spiral
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]:
```suggestion
You can use the `Spiral\Core\ContainerScopeInterface->runScoped` method to achieve this. The default Spiral
```

File: docs/en/framework/scopes.md [hunk 8]
```diff
@@ -1,50 +1,208 @@
...
         dump($container->get(UserContext::class);
-
-    }
+    },
+    bindings: [UserContext::class => $user,],
 );
 ```
 
-The first argument of the `runScope` method is an array of values that should be bound to the container within the 
+The first argument of the `runScope` method is an array of values that should be bound to the container within the
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]:
```suggestion
The second argument of the `runScope` method is an array of values that should be bound to the container within the
```

File: docs/en/framework/scopes.md [hunk 9]
```diff
@@ -1,50 +1,208 @@
...
 callback's scope.
 
-The callback function receives the container as its argument and can use the container to retrieve the values that were 
-bound to it within the callback's scope. This can be useful for creating a temporary scope within which certain values 
+The callback function receives the container as its argument and can use the container to retrieve the values that were
+bound to it within the callback's scope. This can be useful for creating a temporary scope within which certain values
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]: The closure will be autowired. There any parameters can be. The ContainerInterface will resolved as a container of the new scope.

File: docs/en/framework/scopes.md [hunk 10]
```diff
@@ -1,50 +1,208 @@
...
 are available.
 
-In this example, the `runScope` method is creating a temporary scope within which the `UserContext` class is bound to the 
-`$user` variable. The callback function can then use the container to retrieve the `UserContext` object instance, which 
-is available within the callback's scope.
+In this example, the `runScope` method is creating a temporary scope within which the `UserContext` class is bound to
+the `$user` variable. The callback function can then use the container to retrieve the `UserContext` object instance,
+which is available within the callback's scope.
```
Comment (roxblnfk) [inline docs/en/framework/scopes.md]:
```suggestion
In this example, the `runScoped` method is creating a temporary scope within which the `UserContext` class is bound to
the `$user` variable. The callback function can then use the container to retrieve the `UserContext` object instance,
which is available within the callback's scope.
```

File: docs/en/framework/container.md [hunk 11]
```diff
@@ -532,10 +766,106 @@ $invoker->invoke(
...
+final class UserService
+{
+    public function store(User $user): void
+    {
+        //...
+    }
+}
+```
+
+### Scope
```
Comment (wolfy-j) [inline docs/en/framework/container.md:804]: do we have practical examples how it works?

**Review by wolfy-j — APPROVED** [review]

**Review by wolfy-j — APPROVED** [review]

File: docs/en/framework/container.md [hunk 12]
```diff
@@ -268,11 +268,242 @@ public function boot(Container $container): void
...
+));
+```
+
+By setting the `singleton` argument to `true`. This means that whenever you request the time value from the
+container `$container->get('time')`, it will return the same value across multiple invocations.
+
+### DeferredFactory
+
+The `Spiral\Core\Config\DeferredFactory` binding enables you to bind deferred factories to the container using array
+callables.
```
Comment (roxblnfk) [inline docs/en/framework/container.md]:
```suggestion
The `Spiral\Core\Config\DeferredFactory` binding enables you to bind deferred factory to the container using a
special array callable.
```

File: docs/en/framework/container.md [hunk 13]
```diff
@@ -268,11 +268,242 @@ public function boot(Container $container): void
...
+
+$container->bind('some-binding', new DeferredFactory(
+    factory: [MyClass::class, 'handle'],
+));
+```
+
+In the example above, we bind the key some-binding to a `DeferredFactory` instance. The factory property is set
+to `[MyClass::class, 'handle']`. When the some-binding key is requested from the
+container `$container->get('some-binding')`, the `DeferredFactory` will create an instance of `MyClass` and then invoke
+the handle method on that instance.
```
Comment (roxblnfk) [inline docs/en/framework/container.md]:
```suggestion
In the example above, we bind the key some-binding to a `DeferredFactory` instance. The factory property is set
to `[MyClass::class, 'handle']`. When the some-binding key is requested from the
container `$container->get('some-binding')`, the `DeferredFactory` will resolve an instance of `MyClass` and then invoke
the handle method on that instance to produce the required value.
```

File: docs/en/framework/container.md [hunk 14]
```diff
@@ -268,11 +268,242 @@ public function boot(Container $container): void
...
+        $obj->setLogger($logger);
+
+        return $obj;
+    }
+));
+```
+
+By configuring the Inflector binding, you can apply modifications or injections to objects of a specific type
+automatically whenever they are requested from the container. In this case, any object
+implementing `LoggerAwareInterface` will have its logger set based on the specified configuration.
```
Comment (roxblnfk) [inline docs/en/framework/container.md]:
```suggestion
In this case, any object implementing `LoggerAwareInterface`
will have its logger set based on the specified configuration.
```

File: docs/en/framework/container.md [hunk 15]
```diff
@@ -268,11 +268,242 @@ public function boot(Container $container): void
...
+It's important to note that with the `Shared` binding, custom arguments cannot be provided during subsequent requests.
+The object will be created with the initial arguments and reused as is.
+
+It is particularly useful when you want to ensure that the same instance of an object is used throughout the
+application. It provides persistence and prevents the creation of multiple instances with different arguments.
+
+### Inflector
+
+An inflector allows you to manipulate an object before returning it from the container. This is particularly useful for
+applying common modifications or injections to objects of a specific type.
```
Comment (roxblnfk) [inline docs/en/framework/container.md]:
```suggestion
An inflector allows you to manipulate an object after creating it in the container. This is particularly useful for
applying common modifications or injections to objects of a specific type.
```

**Pull request merged**